vllm/model_executor/layers/quantization/bitsandbytes.py

```diff
@@ -0,0 +1,175 @@
+                output_partition_sizes) % quant_ratio != 0:
+            raise ValueError(
+                "The input size is not aligned with the quantized "
+                "weight shape. ")
+        qweight = Parameter(
+            torch.empty(
+                input_size_per_partition * sum(output_partition_sizes) //
+                quant_ratio,
+                1,
+                dtype=torch.uint8,
+            ),
+            requires_grad=False,
+        )
+
+        set_weight_attrs(
+            qweight,
+            {
+                "input_dim": 0,
+                # different from other quantization methods, a tensor of shape [n,m]
+                # is qunatized to [n*m, 1] in BitsAndBytes,
+                # so the output_dim is 0
+                "output_dim": 0,
+                "bnb_quant_state": None,
+                "pack_factor": quant_ratio,
+                "use_bitsandbytes": True,
+            })
+        layer.register_parameter("qweight", qweight)
+        set_weight_attrs(qweight, extra_weight_attrs)
+
+    def apply(self,
+              layer: torch.nn.Module,
+              x: torch.Tensor,
+              bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+
+        orginal_type = x.dtype
+        bf_x = x.to(torch.bfloat16)
+
+        qweight = layer.qweight
+
+        quant_states = getattr(qweight, "bnb_quant_state", None)
+        if quant_states is None:
+            raise RuntimeError("Quantization state is not found.")
+
+        # check quant_states is dictionary of integer to QuantState
+        if not isinstance(quant_states, dict):
+            raise RuntimeError("Quantization state is not a dictionary.")
+        if not all(isinstance(k, int) for k in quant_states.keys()):
+            raise RuntimeError("Quantization state keys are not integers.")
+        if not all(isinstance(v, QuantState) for v in quant_states.values()):
+            raise RuntimeError("Quantization state values are not QuantState.")
+
+        pack_ratio = getattr(qweight, "pack_factor", None)
+        if pack_ratio is None:
+            raise RuntimeError("Pack ratio is not found.")
+        if not isinstance(pack_ratio, int):
+            raise RuntimeError("Pack ratio is not an integer.")
+
+        num_elements = [0] * len(quant_states)
+        for seq, quant_state in enumerate(quant_states.items()):
+            num_elements[seq] = math.prod(quant_state[1].shape) // pack_ratio
+
+        offsets = np.concatenate(([0], np.cumsum(num_elements)))
+
+        out = matmul_4bit(bf_x, qweight[offsets[0]:offsets[1]].t(),
+                          quant_states[0])
+
+        for i in range(1, len(quant_states)):
+
+            current_basic_output = matmul_4bit(
+                bf_x, qweight[offsets[i]:offsets[i + 1]].t(), quant_states[i])
+
+            out = torch.cat((out, current_basic_output), dim=-1)
+
+        out = out.to(orginal_type)
+
+        if bias is not None:
+            out = out + bias
```

Comment:
```suggestion
            out += bias
```
this way we do an in-place addition which saves memory

Reply:
Thanks for catching this.

Reply:
updated.